### PR TITLE
Relative-lr main merge

### PR DESCRIPTION
**Pull request by j321m**
# Description
extracted relative lr functionalities from ludziej's schizo-branch to main.

# Checklist
   - [ ] Is this PR focused on a single feature without the possibility of being divided into smaller PRs?
   - [ ] Are all variables set with logical defaults that are backward compatible?
   - [ ] I attached link to neptune if it was necessary
   - [ ] Have you successfully executed the linter and tests?
   - [ ] Are all functions concise and don't require splitting?
   - [ ] Is there documentation for arguments and complex logic?
   - [ ] Are there no temporary solutions in the code? Should a task be added to Trello to refine the code?
   - [ ] Are all functions placed in appropriate files?
   - [ ] Is the PR name meaningful and descriptive?
   - [ ] Is PR description provided, or unnecessary?
  
Other questions:
   - [ ] If I made changes to the requirements or anything related to the Singularity image I did generate new image and upload a new image to the clusters
   - [ ] If this change is significant enough I notified all individuals working with this repository (@channel in slack channel)
   - [ ] If there is a need for a second reviewer I requested additional review 

# Reviewer's checklist:
   - [ ] Does the Neptune link work and does it actually compare the situation before and after the PR?
   - [ ] Do I understand the code?
   - [ ] Do I think another reviewer is not needed or I commented that this PR need another reviewer?

### CODE REVIEW

File: lizrd/train/scheduler.py [hunk 1]
```diff
@@ -6,15 +6,20 @@
...
 
 def get_scheduler(
-    args, ratios_in_group_order: Optional[list[float]] = None
+    args,
+    ratios_lr_in_group_order: Optional[list[float]] = None,
+    scheduler_fractions_in_group_order: Optional[list[float]] = None,
 ) -> "AbstractLRScheduler":
-    if ratios_in_group_order is None:
-        ratios_in_group_order = [1.0]
+    print(args.scheduler)
```
Comment (maciek-pioro) [inline lizrd/train/scheduler.py:13]: remov

File: research/conditional/utils/argparse.py [hunk 2]
```diff
@@ -364,6 +364,19 @@ def introduce_parser_arguments(
...
+        "--relative_scheduler_fraction",
+        type=load_dict_in_args,
+        default=None,
+        help="""Dictionary with relative final learning rate fractions for different modules
+        Example: --relative_scheduler_fraction "{'attention': 0.1, 'feedforward': 0.1, 'moe': 0.1}
+        Example in config yaml:
+        relative_scheduler_fraction:
+            attention: 0.1
+            feedforward: 0.1
+            moe: 0.1
```
Comment (maciek-pioro) [inline research/conditional/utils/argparse.py:377]: for `relative_lr` change `help` so that it says, that it's the starting relative lr.

File: research/conditional/utils/argparse.py [hunk 3]
```diff
@@ -364,6 +364,19 @@ def introduce_parser_arguments(
...
         """,
     )
+    parser.add_argument(
+        "--relative_scheduler_fraction",
+        type=load_dict_in_args,
+        default=None,
+        help="""Dictionary with relative final learning rate fractions for different modules
+        Example: --relative_scheduler_fraction "{'attention': 0.1, 'feedforward': 0.1, 'moe': 0.1}
+        Example in config yaml:
+        relative_scheduler_fraction:
```
Comment (maciek-pioro) [inline research/conditional/utils/argparse.py:374]: Change help here so it says "relative_lr" is the starting ratio.

File: lizrd/train/scheduler.py (hunk 1 above)
Comment (maciek-pioro) [inline lizrd/train/scheduler.py:13]: Either remove the print or add something like f"schedulers: {args.scheduler}"

File: research/conditional/train/cc_train.py [hunk 4]
```diff
@@ -280,7 +353,17 @@ def main(
...
-    param_grops, ratios_in_group_order = make_param_groups_and_lr_ratios(args, model)
+    # set args for optimizer param_groups
+    baseline_args = [args.learning_rate, args.final_lr_fraction]
+    baseline_keys = [
+        "ratios_lr_in_group_order",
+        "scheduler_fractions_in_group_order",
+    ]  # key names like arguments in get_scheduler :), best idea i had :(
+    relative_params_all = [args.relative_lr, args.relative_scheduler_fraction]
+
+    param_grops, ratios_in_group_order = make_param_groups_for_optimizer(
```
Comment (maciek-pioro) [inline research/conditional/train/cc_train.py:364]: param_groups (typo)

File: research/conditional/train/cc_train.py [hunk 5]
```diff
@@ -280,7 +353,17 @@ def main(
...
             print(name, param.shape)
 
-    param_grops, ratios_in_group_order = make_param_groups_and_lr_ratios(args, model)
+    # set args for optimizer param_groups
+    baseline_args = [args.learning_rate, args.final_lr_fraction]
+    baseline_keys = [
+        "ratios_lr_in_group_order",
+        "scheduler_fractions_in_group_order",
+    ]  # key names like arguments in get_scheduler :), best idea i had :(
+    relative_params_all = [args.relative_lr, args.relative_scheduler_fraction]
```
Comment (maciek-pioro) [inline research/conditional/train/cc_train.py:362]: nie no w sumie plz przezwijmy to relative_scheduler_fraction bo ten kod jest niezrozumiały.
Dosłownie coś w stylu "final_lr_fraction"

File: research/conditional/train/cc_train.py [hunk 6]
```diff
@@ -93,6 +93,79 @@ def make_param_groups_and_lr_ratios(args, model):
     return param_grops, ratios_in_group_order
 
 
+# it's make_param_groups_and_lr_ratios function, but in a for-loop, and it returns param_groups,
+# where each group has a unique combination of relative optimizer related parameters.
+def make_param_groups_for_optimizer(
```
Comment (crewtool) [inline research/conditional/train/cc_train.py:98]: Why this function is in cc_train?

File: research/conditional/train/cc_train.py [hunk 7]
```diff
@@ -93,6 +93,79 @@ def make_param_groups_and_lr_ratios(args, model):
...
+        param_groups = merge_params_dicts(param_groups, group2)
+
+    for param_group in param_groups:
+        for baseline_arg, key in zip(baseline_args, baseline_keys):
+            ratios_in_group_order[key].append(param_group[key] / baseline_arg)
+
+    return param_groups, ratios_in_group_order
+
+
+def merge_params_dicts(param_group_1: list, param_group_2: list):
```
Comment (crewtool) [inline research/conditional/train/cc_train.py:153]: Probably should be moved somewhere else

File: research/conditional/train/cc_train.py (hunk 5 above)
Comment (j321m) [reply research/conditional/train/cc_train.py]: done

File: research/token_reduction/configs/rnn_base_seqlen1024.yaml
Comment (maciek-pioro) [inline research/token_reduction/configs/rnn_base_seqlen1024.yaml:1]: what this do here

File: research/token_reduction/configs/run_mini_on_laptop.yaml
Comment (maciek-pioro) [inline research/token_reduction/configs/run_mini_on_laptop.yaml:1]: what this do here
